File: combine_multiplexed.py
import numpy as np

def combine_multiplexed(multiplexed_data):
    """
    Combine 1024x16x4 multiplexed records into full 1024x61 data.

    Parameters
    ----------
    multiplexed_data : numpy.ndarray
        Expected shape: (num_depths, 16, 4)

    Returns
    -------
    full_data : numpy.ndarray
        Shape: (num_depths, 61)
    """

    # Validate shape
    if multiplexed_data.ndim != 3:
        raise ValueError("Input must be a 3D array")

    num_depths, num_records, num_sections = multiplexed_data.shape

    if num_records != 16 or num_sections != 4:
        raise ValueError("Incorrect input size. Expected shape (N, 16, 4)")

    # Allocate output array
    full_data = np.zeros((num_depths, 61))

    # Transfer data
    for rec in range(16):          # 0–15
        for sec in range(4):       # 0–3
            ch = sec * 16 + rec    # channel index (0-based now)
            if ch < 61:            # only 61 channels
                full_data[:, ch] = multiplexed_data[:, rec, sec]

    return full_data

File: combine_multiplexed.m
% Combine 1024x16x4 multiplexed records into full data (1024x61)
function full_data = combine_multiplexed(multiplexed_data)
[num_depths, num_records, num_sections] = size(multiplexed_data); % Expected : [1024, 16, 4]

% Validate input size
if num_records ~= 16 || num_sections ~= 4
    error('Incorrect input size');
end

% Allocate output array
full_data = zeros(num_depths, 61);

%% Transfer data into a single, 2D output array
for rec = 1:16 % Loop over the 16 records
    for sec = 1:4 % Look at each section
        ch = (sec-1)*16 + rec; % Get channel index
        if ch <= 61 % Only 61 channels
            full_data(:, ch) = multiplexed_data(:, rec, sec); % Transfer all 1024 depth data
        end
    end
end

end